listenbrainz/db/couchdb.py

```diff
@@ -0,0 +1,249 @@
+from sentry_sdk import start_span
+
+from urllib3 import Retry
+
+# stat type followed by a underscore followed by a date in YYYYMMDD format
+DATABASE_NAME_PATTERN = re.compile(r"([a-zA-Z]+)_(\w+)_(\d{8})")
+
+DATABASE_LOCK_FILE = "LOCK"
+
+_user = None
+_admin_key = None
+_host = None
+_port = None
+
+
+def init(user, password, host, port):
+    """
+    Initialize config to connect to couchdb instance.
+    
+    Args:
+        user: couchdb admin user name
+        password: couchdb admin password
+        host: couchdb service host
+        port: couchdb service port
+    """
+    global _user, _admin_key, _host, _port
+    _user = user
+    _admin_key = password
+    _host = host
+    _port = port
+
+
+def get_base_url():
+    return f"http://{_user}:{_admin_key}@{_host}:{_port}"
+
+
+def create_database(database: str):
+    """ Create a couchdb database with the given name.
+
+    For example, if prefix is artists_weekly and the day is 2022-07-10 then the newly
+    created couchdb database will be named artists_weekly_20220710.
+
+    Args:
+         database: the database's name
+    """
+    databases_url = f"{get_base_url()}/{database}"
+    response = requests.put(databases_url)
+    response.raise_for_status()
+
+
+def list_databases(prefix: str) -> list[str]:
+    """ List all couchdb database whose name starts with the given prefix
+    sorted in the descending order of creation.
+
+    Consider statistics, we generate those daily and create a new database each time for each
+    stat daily. We name databases as `prefix_YYYYMMDD` where prefix describes the stat name and
+    YYYYMMDD is the date. After statistics for the day have been inserted, we want to get rid
+    of the older database for that stat. This method looks up all the databases whose name starts
+    with the given prefix.
+    """
+    databases_url = f"{get_base_url()}/_all_dbs"
+    response = requests.get(databases_url)
+    response.raise_for_status()
+    all_databases = response.json()
+
+    databases = [database for database in all_databases if database.startswith(prefix)]
+    databases.sort(reverse=True)
+    return databases
+
+
+def delete_database(prefix: str):
+    """ Delete all but the latest database whose name starts with the given prefix.
+
+    Before deleting, the existence of a LOCK file is checked. If a file named LOCK,
+    exists in the database then it is not deleted.
+
+    Args:
+         prefix: the string to match database names with
+
+    Returns:
+        tuple of name of databases that were deleted and which matched the prefix
+        but weren't deleted
+    """
+    databases = list_databases(prefix)
+    # remove the latest database from the list then delete the databases remaining in the list.
+    databases.pop(0)
+
+    deleted, retained = [], []
+
+    for database in databases:
+        databases_url = f"{get_base_url()}/{database}"
+        response = requests.get(f"{databases_url}/{DATABASE_LOCK_FILE}")
```

Comment:
I think for clarity we should have a check_database_lock() function, rather than having the lock check embedded here.

Reply:
done.